### PR TITLE
Render JS using the velocity engine

### PR DESCRIPTION
**Pull request by mpkorstanje**
It is preferable to render js through the template engine over rendering it at in java. This ensures that any transformations happen as late as possible

### CODE REVIEW

**Comment by mpkorstanje** [comment]
Please think about renaming the other javascript files to .js.vm. I had not looked at the javascript files because I assumed to be static and not velocity templates.

File: src/main/resources/templates/js/trends-chart.js.vm [hunk 1]
```diff
@@ -1,20 +1,23 @@
+#macro(stringArray $array) [#foreach($element in $array) "$element", #end] #end
+#macro(numberArray $array) [#foreach($element in $array) $element, #end] #end
+
```
Comment (mpkorstanje) [inline src/main/resources/templates/js/trends-chart.js.vm]: Trailing comma's in lists are okay in javascript.

File: src/main/resources/templates/generators/overviewTrends.vm [hunk 2]
```diff
@@ -5,7 +5,7 @@
   #includeTitle("Trends Overview")
 
   <script type="text/javascript">
-  #parse("/templates/js/trends-chart.js")
+  #parse("/templates/js/trends-chart.js.vm")
   </script>
```
Comment (mpkorstanje) [inline src/main/resources/templates/generators/overviewTrends.vm]: Renamed the file for clarity - it should be explicit that this is a template.

File: src/main/resources/templates/js/trends-chart.js.vm [hunk 3]
```diff
@@ -1,20 +1,23 @@
...
-    var pendingSteps = $pendingSteps;
-    var undefinedSteps = $undefinedSteps;
+    var passedSteps = #numberArray($passedSteps);
+    var failedSteps = #numberArray($failedSteps);
+    var skippedSteps = #numberArray($skippedSteps);
+    var pendingSteps = #numberArray($pendingSteps);
+    var undefinedSteps = #numberArray($undefinedSteps);
 
-    var durations = $durations;
+    var durations = #numberArray($durations);
```
Comment (mpkorstanje) [inline src/main/resources/templates/js/trends-chart.js.vm]: number array and string array are not strictly needed, javascript will coerce  strings that represent numbers into numbers when needed but this just feels better, more precise.

**Comment by damianszczepanik** [comment]
It does not solve the main regression introduced in pull #576

File: src/main/resources/templates/generators/overviewTrends.vm [hunk 4]
```diff
@@ -5,7 +5,7 @@
   #includeTitle("Trends Overview")
 
   <script type="text/javascript">
-  #parse("/templates/js/trends-chart.js")
+  #parse("/templates/js/trends-chart.js.vm")
```
Comment (damianszczepanik) [inline src/main/resources/templates/generators/overviewTrends.vm]: I don't like this change because of:
- now some files are js, others vm
- by this change IDEA does not color syntax by default

File: src/main/resources/templates/js/trends-chart.js.vm (hunk 1 above)
Comment (damianszczepanik) [inline src/main/resources/templates/js/trends-chart.js.vm]: I prefer to keep logic in Java because it allows me to test it easily and by having more logic in velocity I'm increasing potential regression like the one with tags because this macro is not tested

**Comment by mpkorstanje** [comment]
Works for me:

works for me

Can you verify that you get this as the output:
overview-trends html

File: src/main/resources/templates/generators/overviewTrends.vm (hunk 4 above)
Comment (mpkorstanje) [inline src/main/resources/templates/generators/overviewTrends.vm]: 1. You should also rename the other js files where they are used as templates but that is outside the scope of this change. 

2. Hate to do this. But it works for me:

trends-charts-js vm

File: src/main/resources/templates/js/trends-chart.js.vm (hunk 1 above)
Comment (mpkorstanje) [inline src/main/resources/templates/js/trends-chart.js.vm]: That is architecturally not a sensible decision. The template engine is responsible for transforming the data into the output format and knows what context it will be injected into. Sidestepping this by using java adds complexity as java does not know the output format or any restrictions that apply such as the need to escape characters.

I will look into a test to see if the correct lines are output into overview-trends.html but frankly said, this part of the code is not setup for automated testing and will always need a manual test.

**Comment by mpkorstanje** [comment]
Additional test added to TrendsOverviewPageIntegrationTest

**Comment by mpkorstanje** [comment]
Found more manual array creation in the TagsOverviewPage.

**Comment by mpkorstanje** [comment]
TagsOverviewPage is fixed.

works2

File: src/main/resources/templates/js/tags-chart.js.vm [hunk 5]
```diff
@@ -1,32 +1,35 @@
+#macro(stringArray $array) [#foreach($element in $array) "$element", #end] #end
+#macro(numberArray $array) [#foreach($element in $array) $element, #end] #end
+
```
Comment (mpkorstanje) [inline src/main/resources/templates/js/tags-chart.js.vm]: Duplicated this. Best way to dedupe it?

File: src/main/java/net/masterthought/cucumber/generators/TagsOverviewPage.java [hunk 6]
```diff
@@ -38,14 +36,14 @@ public void prepareReport() {
...
     }
 
-    static String generateTagLabels(List<TagObject> tagsObjectList) {
+    static String[] generateTagLabels(List<TagObject> tagsObjectList) {
         int tagCount = tagsObjectList.size();
         String[] tagNames = new String[tagCount];
 
         for (int i = 0; i < tagCount; i++) {
-            tagNames[i] = StringUtils.wrap(tagsObjectList.get(i).getName(), "\"");
+            tagNames[i] = tagsObjectList.get(i).getName();
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/cucumber/generators/TagsOverviewPage.java]: so now you have some logic here and also in macro.js.vm file what is even worse than now :(

File: src/main/resources/templates/generators/overviewTags.vm [hunk 7]
```diff
@@ -5,7 +5,7 @@
   #includeTitle("Tags Overview")
 
   <script type="text/javascript">
-  #parse("/templates/js/tags-chart.js")
+  #parse("/templates/js/tags-chart.js.vm")
```
Comment (damianszczepanik) [inline src/main/resources/templates/generators/overviewTags.vm]: if you want to change it to vm then apply this change for all files - after your change some JavaScript charts are in js files, others in vm

File: src/main/resources/templates/macros/array.js.vm [hunk 8]
```diff
@@ -0,0 +1,2 @@
+#macro(stringArray $array) [#foreach($element in $array) "$element", #end] #end
```
Comment (damianszczepanik) [inline src/main/resources/templates/macros/array.js.vm]: name of the file tells developer that there is some JavaScript code here - there is no

File: src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java [hunk 9]
```diff
@@ -66,13 +66,14 @@ public void prepareReport_setTagsToExcludeFromChart_ReturnsFilteredTags() {
...
         // then
         VelocityContext context = page.context;
-        assertThat(context.get("chart_categories")).isEqualTo("[\"@fast\"]");
-        assertThat(context.get("chart_data")).isEqualTo(asList(
-                "[100.00]",
-                "[0.00]",
-                "[0.00]",
-                "[0.00]",
-                "[0.00]"));
+        assertThat(context.get("chart_categories")).isEqualTo(new String[]{"@fast"});
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java]: there is probably method to extract element from that array and then compare the first one so `new String[]` is not needed any more

File: src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java [hunk 10]
```diff
@@ -66,13 +66,14 @@ public void prepareReport_setTagsToExcludeFromChart_ReturnsFilteredTags() {
...
+        assertThat(context.get("chart_data")).isEqualTo(new String[][]{
+                {"100.00"},
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java]: same

File: src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java [hunk 11]
```diff
@@ -82,10 +83,10 @@ public void generateTagLabels_ReturnsTags() {
         List<TagObject> allTags = this.tags;
 
         // when
-        String labels = TagsOverviewPage.generateTagLabels(allTags);
+        String[] labels = TagsOverviewPage.generateTagLabels(allTags);
 
         // then
-        assertThat(labels).isEqualTo("[\"@checkout\",\"@fast\",\"@featureTag\"]");
+        assertThat(labels).isEqualTo(new String[]{"@checkout","@fast","@featureTag"});
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java]: same

File: src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java [hunk 12]
```diff
@@ -95,14 +96,15 @@ public void generateTagValues_ReturnsTagValues() {
...
+        String[][] labels = TagsOverviewPage.generateTagValues(allTags);
 
         // then
-        assertThat(labels).containsExactly(
-                "[62.50, 100.00, 100.00]",
-                "[6.25, 0.00, 0.00]",
-                "[12.50, 0.00, 0.00]",
-                "[6.25, 0.00, 0.00]",
-                "[12.50, 0.00, 0.00]");
+        assertThat(labels).isEqualTo(new String[][]{
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java]: I believe there is some containsExactly for two dimentional arrays - take a look if it won't be simpler

File: src/test/java/net/masterthought/cucumber/generators/TrendsOverviewPageTest.java [hunk 13]
```diff
@@ -66,61 +66,19 @@ public void prepareReport_AddsCustomProperties() {
         VelocityContext context = page.context;
         assertThat(context.getKeys()).hasSize(18);
 
-        assertThat(context.get("buildNumbers")).isEqualTo("[\"01_first\",\"other build\",\"05last\"]");
-        assertThat(context.get("failedFeatures")).isEqualTo("[1,2,5]");
-        assertThat(context.get("totalFeatures")).isEqualTo("[10,20,30]");
+        assertThat(context.get("buildNumbers")).isEqualTo(new String[]{"01_first","other build","05last"});
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/TrendsOverviewPageTest.java]: same

File: src/test/java/net/masterthought/cucumber/generators/integrations/TagsOverviewPageIntegrationTest.java [hunk 14]
```diff
@@ -70,6 +70,30 @@ public void generatePage_generatesCharts() {
...
+        // check that data used by the charts is correctly inserted into the script section
+
+        assertThat(document.html()).contains("labels:  [ \"@checkout\",  \"@fast\",  \"@featureTag\", ]");
+        assertThat(document.html()).contains("data:  [ 62.50,  100.00,  100.00, ]");
+        assertThat(document.html()).contains("data:  [ 6.25,  0.00,  0.00, ]");
+        assertThat(document.html()).contains("data:  [ 12.50,  0.00,  0.00, ]");
+        assertThat(document.html()).contains("data:  [ 6.25,  0.00,  0.00, ]");
+        assertThat(document.html()).contains("data:  [ 12.50,  0.00,  0.00, ]");
+
+    }
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/integrations/TagsOverviewPageIntegrationTest.java]: nice

File: src/test/java/net/masterthought/cucumber/generators/integrations/TrendsOverviewPageIntegrationTest.java [hunk 15]
```diff
@@ -75,4 +75,33 @@ public void generatePage_generatesCharts() {
...
+        assertThat(document.html()).contains("var totalFeatures =  [ 10,  20,  30, ] ;");
+        assertThat(document.html()).contains("var failedScenarios =  [ 10,  20,  20, ] ;");
+        assertThat(document.html()).contains("var totalScenarios =  [ 10,  2,  5, ] ;");
+        assertThat(document.html()).contains("var passedSteps =  [ 1,  3,  5, ] ;");
+        assertThat(document.html()).contains("var failedSteps =  [ 10,  30,  50, ] ;");
+        assertThat(document.html()).contains("var skippedSteps =  [ 100,  300,  500, ] ;");
+        assertThat(document.html()).contains("var pendingSteps =  [ 1000,  3000,  5000, ] ;");
+        assertThat(document.html()).contains("var undefinedSteps =  [ 10000,  30000,  50000, ] ;");
+        assertThat(document.html()).contains("var durations =  [ 3206126182398,  3206126182399,  3206126182310, ] ;");
+
```
Comment (damianszczepanik) [inline src/test/java/net/masterthought/cucumber/generators/integrations/TrendsOverviewPageIntegrationTest.java]: very nice

File: src/main/resources/templates/macros/array.js.vm (hunk 8 above)
Comment (mpkorstanje) [inline src/main/resources/templates/macros/array.js.vm]: This will render 

    ["something", "something", "something",] 

which is a javascript array. Unsure what I should call it.

File: src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java (hunk 9 above)
Comment (mpkorstanje) [inline src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java]: There is IntArrayAssert that allows these operations to work on arrays but is these are created through: `assertThat(int[] actual)`. Because `context.get()` always returns an object, the method binds to  `assertThat(T actual)` and we don't get that fancy matcher. .

Comment (mpkorstanje) [inline src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java]: Can do a cast and then use containsExactly but that doesn't improve readability.  Example:

```java
        assertThat((String[]) context.get("buildNumbers")).containsExactly("01_first", "other build", "05last");
        assertThat((int[]) context.get("failedFeatures")).containsExactly(1, 2, 5);
        assertThat((int[]) context.get("totalFeatures")).containsExactly(10, 20, 30);

        assertThat((int[]) context.get("failedScenarios")).containsExactly(10, 20, 20);
        assertThat((int[]) context.get("totalScenarios")).containsExactly(10, 2, 5);

        assertThat((int[]) context.get("passedSteps")).containsExactly(1, 3, 5);
        assertThat((int[]) context.get("failedSteps")).containsExactly(10, 30, 50);
        assertThat((int[]) context.get("skippedSteps")).containsExactly(100, 300, 500);
        assertThat((int[]) context.get("pendingSteps")).containsExactly(1000, 3000, 5000);
        assertThat((int[]) context.get("undefinedSteps")).containsExactly(10000, 30000, 50000);

        assertThat((long[]) context.get("durations")).containsExactly(3206126182398L, 3206126182399L, 3206126182310L);
```

File: src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java (hunk 12 above)
Comment (mpkorstanje) [inline src/test/java/net/masterthought/cucumber/generators/TagsOverviewPageTest.java]: Makes things worse

```java
  // then
        assertThat(labels).containsExactly(
                new String[]{"62.50", "100.00", "100.00"},
                new String[]{"6.25", "0.00", "0.00"},
                new String[]{"12.50", "0.00", "0.00"},
                new String[]{"6.25", "0.00", "0.00"},
                new String[]{"12.50", "0.00", "0.00"}
        );
```

File: src/main/java/net/masterthought/cucumber/generators/TagsOverviewPage.java (hunk 6 above)
Comment (mpkorstanje) [inline src/main/java/net/masterthought/cucumber/generators/TagsOverviewPage.java]: Previously the function had two responsibilities. 1) transform the the list of objects into an array of strings and 2) translate that array of strings into a string that represents a javascript array of strings. The latter was also done in two different places in two different ways.  This has been split into two functions and the redundancy has been removed.

Both the transformation and the translation are now tested from end to end which wasn't done before as the tests did not look at the generated javascript. So in all I reckon we're up from where we were.

File: src/main/resources/templates/generators/overviewTags.vm (hunk 7 above)
Comment (mpkorstanje) [inline src/main/resources/templates/generators/overviewTags.vm]: Ok.
